spring-integration-core/src/main/java/org/springframework/integration/support/MessageBuilder.java

```diff
@@ -176,7 +181,16 @@ public T getPayload() {
 	 */
 	@Override
 	public MessageBuilder<T> copyHeadersIfAbsent(Map<String, ?> headersToCopy) {
-		this.headerAccessor.copyHeadersIfAbsent(headersToCopy);
+		if (headersToCopy != null) {
+			for (Map.Entry<String, ?> entry : headersToCopy.entrySet()) {
+				if (!this.headerAccessor.isReadOnly(entry.getKey())) {
+					this.headerAccessor.setHeaderIfAbsent(entry.getKey(), entry.getValue());
+				}
+				else if (logger.isWarnEnabled()) {
+					logger.warn("The header [" + entry + "] is ignored for population because it is is readOnly.");
```

Comment:
Why do you talk only about `GatewayMethodInboundMessageMapper`?
It is fundamental restriction of any `MessageBuilder` usage.

Any `MessageBuilder.copyHeadersIfAbsent()` leads right now to the silent headers dropping...